File: src/test/java/funkies/LocalizedSystemExceptionTest.java
package funkies;

import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.EnumSource;
import org.junit.jupiter.params.provider.MethodSource;

import java.util.stream.Stream;

import static funkies.LocalizedErrorMessages.Lang.EN;
import static funkies.LocalizedErrorMessages.Lang.PL;
import static funkies.errors.UserError.ACCESS_DENIED;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.AssertionsForClassTypes.assertThatCode;

class LocalizedSystemExceptionTest {

    @ParameterizedTest
    @MethodSource
    void localizedSystemExceptionIsThrowable(LocalizedErrorMessages.Lang givenLang, String expectedErrorMessage){
        //when
        var localizedException = LocalizedSystemException.localized(givenLang, ACCESS_DENIED);
        
        //then
        assertThatCode( () -> { throw localizedException; })
                .extracting(LocalizedSystemException.class::cast)
                .satisfies( ex -> {
                    assertThat(ex.getCode()).isEqualTo(ACCESS_DENIED);
                    assertThat(ex.getMessage()).isEqualTo(expectedErrorMessage);
                });
    }

    static Stream<Arguments> localizedSystemExceptionIsThrowable() {
        return Stream.of(
                Arguments.of(EN, "[Localized] Access is denied"),
                Arguments.of(PL, "[Localized] Dostęp zabroniony")
        );
    }
}

File: src/test/kotlin/kfunkies/LocalizedSystemExceptionTest.kt
package kfunkies

import kfunkies.Lang.EN
import kfunkies.Lang.PL
import kfunkies.errors.UserError
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource

class LocalizedSystemExceptionTest {

    @ParameterizedTest
    @MethodSource("localizationErrorTestData")
    fun `Localized exception - same code different messages for lang`(givenLang: Lang, expectedErrorMessage: String) {
        //when
        val localizedException = LocalizedSystemException.localized(
            lang = givenLang, 
            errorCode = UserError.ACCESS_DENIED
        )

        //then
        assertThatThrownBy { throw localizedException }
            .extracting { it as LocalizedSystemException }
            .satisfies(
                { assertThat(it.errorCode).isEqualTo(UserError.ACCESS_DENIED) },
                { assertThat(it.message).isEqualTo(expectedErrorMessage) }
            )
    }

    companion object {
        @JvmStatic
        fun localizationErrorTestData() = listOf(
            EN withResult "[Localized] Access is denied",
            PL withResult "[Localized] Dostęp zabroniony"
        )
    }

}

infix fun Lang.withResult(message : String) = Arguments.of(this, message)  